Fetch permits

zugdev opened the pull request:
Resolves #368

Requires https://github.com/ubiquity-os-marketplace/text-conversation-rewards/issues/284

<!--
- You must link the issue number e.g. "Resolves #1234"
- Please do not replace the keyword "Resolves" https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

At this stage of the PR I have the following set of features:

1. GitHub Authentication
1. Reads from URL if present
1. Fetches permits from Supabase if authenticated and no URL permits
1. Renders the ones which have all the proper data in Supabase

For the permits generated previous to https://github.com/ubiquity-os-marketplace/text-conversation-rewards/issues/284 : I will either write a script to backfill the NULL columns in Supabase, but preferably I will scrape from GitHub issue given that location is present for most permits.

update: I've decided to go with the backfill on the backend so this PR requires https://github.com/ubiquity-os-marketplace/text-conversation-rewards/issues/284.

https://github.com/ubiquity/pay.ubq.fi/pull/383#issuecomment-2670469534

zugdev commented:
As of now, what allows us to fetch `permits` is this RLS rule. 

<img width="1212" alt="image" src="https://github.com/user-attachments/assets/f7bbc3d3-cd0c-46e1-9502-780911041b56" />

However the table `partners`, for example, doesn't have that setup.

<img width="1205" alt="image" src="https://github.com/user-attachments/assets/683f5739-db7f-411d-920c-523802489770" />

In my fork of the DB I will create the same rule, however someone with better context can decide if it's appropriate in production. We also have to decide for table `tokens` but that one seems ok to be public.

zugdev commented:
Yo @EresDev any idea why the workflow fails here?

EresDev commented:
> Yo @EresDev any idea why the `Run unit tests for pages functions / testing` workflow fails here?

Similar error when we had problem with yarn version.  Try generating yarn.lock and committing with correct yarn version.

EresDev commented:
I see your yarn changes locked versions in yarn.lock. Updates them to new allowed versions. This was same problem before. 

Grok's [response](https://x.com/i/grok/share/sNxju15astrNWYuSoCpGt9jjm) seems helpful.

zugdev commented:
1. @0x4007, to demo, I need you to add `https://**.pay-ubq-fi.pages.dev/**` into the Supabase Redirect URL Configuration.

<img width="688" alt="image" src="https://github.com/user-attachments/assets/266c23cd-9e73-4ef8-afe5-22e1d531461a" />

2. @gentlementlegen or @0x4007, I need one of you to run the [backfill permits action](https://github.com/ubiquity-os-marketplace/text-conversation-rewards/actions/workflows/permits-backfill.yml).

0x4007 commented:
It would be prudent to back up database before running the action

zugdev commented:
That's a good idea yes, this PR already works.

Keyrxng commented:
> It would be prudent to back up database before running the action

We have a daily backup running

zugdev commented:
The workflows are dead here for some reason, not sure why yet.

0x4007 commented:
All of them say it's crashing using rpc handler @EresDev can you look into it

zugdev commented:
upgrading rpc-handler to 1.4.0 in permit-generation might do it

EresDev commented:
> All of them say it's crashing using rpc handler @EresDev can you look into it

Looks like it has been resolved here. Link me to other issue if it is failing somewhere else and I will look into it.

zugdev commented:
> > All of them say it's crashing using rpc handler @EresDev can you look into it
> 
> Looks like it has been resolved here. Link me to other issue if it is failing somewhere else and I will look into it. 

It's not on you, nor rpc-handler. It's on how yarn was resolving rpc-handler `1.3.0` and `^1.3.0` within my machine :p

0x4007 reviewed (changes requested):
![image](https://github.com/user-attachments/assets/3d4e272c-8d64-4208-9c04-28506a70eeb3)

Throwing an error so this isn't ready

0x4007 commented:
![image](https://github.com/user-attachments/assets/507d317c-4f4d-44ca-ab6d-9a06ce5e25de)

![image](https://github.com/user-attachments/assets/6598c476-e8d0-40ab-b838-3485c3f9eb7c)

![image](https://github.com/user-attachments/assets/4d6486ef-0a88-4e72-bfce-15e184fa49a3)

One of the three says it was already claimed 

The icons are a total mess when switching between permits. 

This needs a lot of clean up.

zugdev commented:
@0x4007 I see, I think the best COA is:

This PR:
- [ ] Fix the error toasters showing up.
- [ ] Fix already claimed reward showing up.

UI Refac #375:
1. Already solves the icons mess.
1. Clean ups.
1. Show associated GitHub issue when fetching from Supabase.

I believe that to be the best way to procede because adding elements to the table would only render in refactoring it in #375. Potentially they are already solved there.

EresDev commented on static/scripts/rewards/render-transaction/fetch-permits.ts:
```diff
@@ -59,6 +82,15 @@ export async function readClaimDataFromUrl(app: AppState) {
 }
 
+async function readClaimDataFromUrl(): Promise<PermitReward[]> {
+  // safeguard: no claim data found from URL load from supabase
+  if (!base64encodedTxData) {
+    return [];
+  }
```
this check happens before we call this function
```
if (base64encodedTxData) {
    permits = await readClaimDataFromUrl();
  } else if (token) {
```

can we remove this check?

EresDev commented on static/scripts/rewards/render-transaction/render-transaction.ts:
```diff
@@ -1,11 +1,13 @@
 import { insertErc20PermitTableData, insertErc721PermitTableData } from "./insert-table-data";
 import { renderEnsName } from "./render-ens-name";
 import { renderNftSymbol, renderTokenSymbol } from "./render-token-symbol";
+import { TokenType } from "@ubiquibot/permit-generation/types";
```
My vscode wants to import from 
```
import { TokenType } from "@ubiquibot/permit-generation";
```

EresDev commented on static/scripts/rewards/render-transaction/supabase-getters.ts:
```diff
@@ -0,0 +1,155 @@
+
+declare const SUPABASE_URL: string;
+declare const SUPABASE_ANON_KEY: string;
+export const supabase = createClient<Database>(SUPABASE_URL, SUPABASE_ANON_KEY);
```
log if supabase url and key are empty? could be browser console or build console?

zugdev replied:
build should already cover it
<img width="1017" alt="image" src="https://github.com/user-attachments/assets/c1ed1902-8188-48d4-a8d5-96f5d5ef5698" />

but I'll add a runtime check too

EresDev commented on static/scripts/rewards/render-transaction/supabase-getters.ts:
```diff
@@ -0,0 +1,155 @@
+  return data as SupabaseToken;
+}
+
+export async function fetchUserById(userId: number): Promise<SupabaseUser | null> {
+  const { data, error } = await supabase.from("users").select("*").eq("id", userId).single();
+  if (error) {
```
did you consider using graphql for supabase? are we using graphql in any other repository for supabase? A graphql query would have done these many requests in one go.

zugdev replied:
not really, we don't use it in text-conversation-rewards. it's in beta, so I am not sure, however I am experimenting with it now.

EresDev commented on static/scripts/rewards/render-transaction/supabase-getters.ts:
```diff
@@ -0,0 +1,155 @@
+}
+
+export async function fetchPermitsFromSupabase(userId: number): Promise<PermitReward[]> {
+  const { data, error } = await supabase.from("permits").select("*").eq("beneficiary_id", userId);
```
add order by.  For me permits were mixed with claimed permits, and it appeared to be a random order. Show newer first maybe?

EresDev commented on static/scripts/rewards/render-transaction/supabase-getters.ts:
```diff
@@ -0,0 +1,155 @@
+    } else {
+      const oldNonce = BigNumber.from(current.nonce);
+      const newNonce = BigNumber.from(permit.nonce);
+      if (newNonce.gt(oldNonce)) {
```
> Be sure to check for doubles that have the same nonce. This means that the permits were generated multiple times for the same issue. Pick the higher value one if there are two of the same nonce.

Is this what we are trying to do here? If so, you need to compare the amount, not nonce. Nonce would be same. 

Also, when would signature be same? I am lacking this knowledge. My idea is signature would always be different unless everything (amount, receiver, token) is same.

zugdev replied:
signature should never be the same, you are right: it's the hash of the entire object. this was wrong, thanks for catching it. I don't see duplicate nonce's in the DB tho -- anyways it's a great safeguard. also making a map prevents duplicate entries (even with same signature and nonce to be included)

EresDev replied:
- this issue had 2 permits generated for me. https://github.com/ubiquity/rpc-handler/issues/60#issuecomment-2704207649
One 23.93 and other100. By specs, it should show 100 one. But it is showing 23.93 one and ignoring 100 one.

![image](https://github.com/user-attachments/assets/8d6a6c8d-6562-4e30-8807-83b9115fff93)

zugdev replied:
@EresDev can you please check console for `filtered permits` so we can better understand what happened here.

<img width="728" alt="image" src="https://github.com/user-attachments/assets/dd815636-d563-4f8c-9298-c9138339b9cc" />

also try with latest version cause I changed it since this comment

0x4007 replied:
> - this issue had 2 permits generated for me. https://github.com/ubiquity/rpc-handler/issues/60#issuecomment-2704207649
> 
> One 23.93 and other100. By specs, it should show 100 one. But it is showing 23.93 one and ignoring 100 one.
> 
> 
> 
> ![image](https://github.com/user-attachments/assets/8d6a6c8d-6562-4e30-8807-83b9115fff93)
> 
> 

Thanks for the detailed QA testing!

EresDev reviewed (changes requested):
- network id error pops up too often. There is network id in permit, and there is network id in wallet connection, and there are situation in your implementation when you neither have permit, nor have wallet connection. Cover all these cases and prevent this error from showing up unnecessarily.
![image](https://github.com/user-attachments/assets/bf82a65f-9b66-4f27-afaf-a18276f3685b)

- reduce error message on toaster but keep complete message in console, when user declines wallet connect request in the wallet.
![image](https://github.com/user-attachments/assets/04b025d6-db1f-4d2f-a390-fd5c55563e5b)

- Login with Github gets behind the permit details
![image](https://github.com/user-attachments/assets/a3def4c3-e94a-43cd-be4e-d9b9b837f901)

- can we hide details table, "Loading...", cards disclaimer when we open page without permit url and hasn't logged into github yet. Not just hide them but also prevent them from loading unnecessarily if possible. 

![image](https://github.com/user-attachments/assets/a9a6ba3e-d55d-49bc-8d51-f7663f3cf300)

zugdev commented:
@EresDev try again and lmk, the only thing I haven't implemented is GraphQL and the reasons are above. basically I didn't see a big performance gain to justify having a beta product and another package installed.

0x4007 commented:
I don't think the deploy links updated because 

![image](https://github.com/user-attachments/assets/8bbf3c8f-2cd8-4530-88ce-e671606539ce)

zugdev commented:
Yes it's broken, I've opened an issue and have a solution ready in a sec.

https://github.com/ubiquity/pay.ubq.fi/issues/387

EresDev commented on static/scripts/rewards/render-transaction/fetch-permits.ts:
```diff
@@ -59,6 +92,10 @@ export async function readClaimDataFromUrl(app: AppState) {
   await renderTransaction();
 }
 
+async function readClaimDataFromUrl(): Promise<PermitReward[]> {
+  return decodeClaimData(base64encodedTxData ?? "");
+}
```
Now the whole functions `readClaimDataFromUrl` look redundant. Can we just  use `decodeClaimData` and rename it to `decodeClaimDataFromUrl`

EresDev commented on static/scripts/rewards/render-transaction/supabase-getters.ts:
```diff
@@ -0,0 +1,158 @@
+import { createClient } from "@supabase/supabase-js";
+import { Database, type PermitReward } from "@ubiquity-os/permit-generation";
+import { TokenType } from "@ubiquibot/permit-generation/types";
```
```suggestion
import { TokenType } from "@ubiquibot/permit-generation";
```

zugdev replied:
this doesn't work for me

<img width="513" alt="image" src="https://github.com/user-attachments/assets/08106d47-e9dc-469f-8930-1edbbf7d6c06" />

<img width="519" alt="image" src="https://github.com/user-attachments/assets/bff8a1be-eab2-45c4-8f13-dbda33418a68" />

EresDev reviewed (changes requested):
- not covered, still long err message

> reduce error message on toaster but keep complete message in console, when user declines wallet connect request in the wallet.

![image](https://github.com/user-attachments/assets/e337cd3d-1696-45fd-92f7-1a35442e9fa3)

EresDev commented:
> I don't think the deploy links updated because

I have manually run the build/deploy. It should be with above recent commit now if you want to test.

zugdev commented:
thanks. 

1. I don't get the stack in my messages. can you please tell me the steps to reproduce?

localhost
<img width="999" alt="image" src="https://github.com/user-attachments/assets/68db38c7-9e5b-41f5-950a-ed8f1e1f0f01" />

https://7e60a118.pay-ubq-fi.pages.dev/
<img width="980" alt="image" src="https://github.com/user-attachments/assets/54ea212c-4c02-48f9-9bc9-bf103cc637fc" />

EresDev commented on static/scripts/rewards/render-transaction/supabase-getters.ts:
```diff
@@ -0,0 +1,158 @@
+
+async function processPermits(permits: SupabasePermit[]): Promise<PermitReward[]> {
+  const processed = await Promise.all(
+    permits.map(async (permit) => {
+      if (!permit.beneficiary_id || !permit.token_id || !permit.partner_id) {
+        // skip if not enough info from db
+        return null;
+      }
+
+      // fetch objects from db
+      const userRecord = await fetchUserById(permit.beneficiary_id);
```
`fetchUserById` and also other fetch functions of tokens and related data runs in permits loop. I have about 15 permits. 6 of which are of our interest (the unclaimed ones). This function run 15 times and requests same data for all permit related tables. However, I also understand this will not always be same data. 

It is taking about 15 second to load permits for me. One way to improve would be to have a cleaner solution to not request same data. But the minimum improvement we should do is to filter unclaimed permits before running this loop so that we don't request unnecessary and duplicate data of already claimed permits. 

Maybe join was a better idea, if not graphql. 

Also, do you know what is "transaction" column in "permits" table? Is it for claimed one?

Keyrxng replied:
Transaction column is the txhash yeah, ~790 permits validated in the below data that I linked previously.

https://github.com/Keyrxng/ubq-.github/blob/refactor/v2/airdrop-cli/src/scripts/data/db-entries.json

![image](https://github.com/user-attachments/assets/d47364d2-9b55-444e-b268-749800f963cf)

---

https://github.com/Keyrxng/ubq-.github/blob/refactor/v2/airdrop-cli/src/scripts/data/dc-unspent-permits.json

This is a JSON of permits which are verified to be unclaimed. Maybe that's a simple solution to your problem Eres, or is there or there-abouts.

zugdev replied:
I am modifying from a concurrent to a fetch-all approach, but there are some concerns to address:

1. It is extremely unlikely that this harms performance, specially cause those requests fire up concurrently and resolve at basically the same time. The bottleneck in this system is the blockchain part of it. 

2. Though the transaction column exists and should be filled to indicate a claimed permit -- there is written code for that -- it isn't in prod table. Therefore, for every permit, we have to check if `isValidNonce()` - as there is no plausible condition to filter on fetch with.

<img width="260" alt="image" src="https://github.com/user-attachments/assets/5573f9f1-313f-41fa-a6a1-e9096efe2830" />

3. I will time every process for a better performance analysis.

EresDev replied:
> Simply filter by the connected wallet, no github login required etc so less friction overall and idk maybe a speed improvement.

I thought about this and I like the idea. But since the issue is already on the path to use github ID, let's see keep it for now.

EresDev reviewed (changes requested):
> I don't get the stack in my messages. can you please tell me the steps to reproduce?

You need to disconnect your metamask, and reload page again. It asks for metamask connection permission, and you decline it. Here is a video


https://github.com/user-attachments/assets/76119af2-d6a2-4fdd-bd2a-b6b2c3eb211c

--- 

> "fixing" the path doesn't work for me

You are right. seems like vscode problem then. 

---

- you missed this one https://github.com/ubiquity/pay.ubq.fi/pull/383#discussion_r1986701559

---

- also convert PR to "read for review" before requesting review, otherwise it feels like happening by mistake.

zugdev commented:
1. Improved error handling for Metamask:

<img width="539" alt="image" src="https://github.com/user-attachments/assets/a651e5d4-3b01-4788-9d33-9290bed8bbe5" />

2. Improved fetching with joins and minimized queries.

<img width="346" alt="image" src="https://github.com/user-attachments/assets/8229a172-a417-4f00-a02e-612492d090c7" />

Bottleneck is on filtering because as of now we check nonce on-chain.

zugdev commented:
@EresDev please read above and inside the changes too.

I will write a proposal to minimize the number of on-chain calls to enhance performance.

0x4007 commented:
Given that we are loading a list of rewards, I wonder if it makes sense to eventually change to our list style view template. 

Basically like work.ubq.fi

0x4007 commented:
Just tested and waiting for 10+ seconds to use the app is unacceptable you'll need to stream the render at least. 

![image](https://github.com/user-attachments/assets/f87f3542-a92e-4940-9913-6bb708ade83c)

![image](https://github.com/user-attachments/assets/9ec0e481-02bf-4ae8-89fb-5e07cc9e0e30)

zugdev commented:
> Given that we are loading a list of rewards, I wonder if it makes sense to eventually change to our list style view template.
> 
> Basically like work.ubq.fi

It does.

> If filtering takes time you need to do caching and stream the render. But also I designed the database schema to have the claim transaction link saved. It was supposed to be POSTed from this client when the user claims.

Filtering takes time because of on-chain calls. I think the problem is that in Supabase permits table there is no RLS policy in place for updates, only for reads.

<img width="1007" alt="image" src="https://github.com/user-attachments/assets/8ccbd207-c518-410c-8ce4-2960b2ce2559" />

We should create that policy for an user. I think it's best you manage that since you have better context.

I will improve render time with in-browser caching though.

zugdev commented:
> Given that we are loading a list of rewards, I wonder if it makes sense to eventually change to our list style view template.
> 
> Basically like work.ubq.fi

It does.

> If filtering takes time you need to do caching and stream the render. But also I designed the database schema to have the claim transaction link saved. It was supposed to be POSTed from this client when the user claims.

Filtering takes time because of on-chain calls. I think the problem is that in Supabase permits table there is no RLS policy in place for updates, only for reads.

<img width="696" alt="image" src="https://github.com/user-attachments/assets/6ba93a6c-5cd1-425b-8313-31d8e2063ee6" />

<img width="1001" alt="image" src="https://github.com/user-attachments/assets/063f2d7e-8c0e-4985-a41e-86b3bc8dcdc6" />

We should create a policy for an user. I think it's best you manage that since you have better context.

I will improve render time with in-browser caching though.

0x4007 commented:
Streaming the permits is the way to speed up render time. Besides, with the current UX you just need to display the first one immediately. 

RLS is not relevant for read speed. It is for writing the transaction link though.

zugdev commented:
I've implemented localstorage caching for claimed permits in `isInvalidNonce`, now cache hits are virtually 0ms. I was thinking about caching unclaimed but this would disrupt the flow between devices a bit - I'd rather have it in the DB whenever we create that rule.

<img width="419" alt="image" src="https://github.com/user-attachments/assets/147c1938-a4c7-48a0-b9c2-4b4bb182e67a" />

About your suspicious number of permits, you can check console for `permits` and it will show you your raw Supabase permits, valid unfiltered and filtered. Also will log specially the ones that were cut out for not having enough data.

<img width="426" alt="image" src="https://github.com/user-attachments/assets/66130989-197b-431f-b64f-fa665a9c8184" />

There were brief times in history where the rewards plugin wasn't reliable and it would either generate ghost permits, or be weird. Here is a personal example of 3 invalid permits:

<img width="400" alt="image" src="https://github.com/user-attachments/assets/22480510-c7e7-4ce1-b610-34b53e6af8a4" />

This location ID refers to https://github.com/ubiquity-os/ubiquity-os-plugin-installer/issues/26. I was never involved in this issue and therefore I don't have a permit there. All of my other invalid permits are of the same nature - and their ids are below 650. I have permits with ids from 480 - 1200 and my invalid ones are around 630 and 570.

zugdev commented:
> Streaming the permits is the way to speed up render time. Besides, with the current UX you just need to display the first one immediately.
> 
> RLS is not relevant for read speed. It is for writing the transaction link though.

RLS in supabase is an auth policy set, it is necessary to create a rule to allow updating the permits table with transaction hash from our webapp. The problem is not read speed.

zugdev commented:
Now streaming permits render.

0x4007 commented:
Will have to test later 

![image](https://github.com/user-attachments/assets/f4833b40-961d-4a9f-8d4d-b00b871415dc)

zugdev commented:
Simulating the experience as @0x4007 

https://github.com/user-attachments/assets/9a3cfef4-eff6-4577-be74-75906bb3fc4e

0x4007 commented:
Just tested it. The first permit should be instantly claimable. While the user is busy claiming it, it buys the app more time to fetch the subsequent one and so on- this yields a seamless user expeirence. This is what I meant by "streaming renders"

EresDev commented:
When permits load and you click claim button, all unclaimed permits requests show up at once in metamask.  Was it intentional? In fact, my permits are 6, and it is showing 7 requests. This doesn't happen when I use prev/next. If go to previous permit, and click claim, it just tries to claim 1 permit. 

![image](https://github.com/user-attachments/assets/ec0cadad-5ed9-48a4-9b3b-5651b2106835)

And if you click the normal cancel button, and not "Reject All". 

![image](https://github.com/user-attachments/assets/52a8bd03-f1da-4de0-b9c6-d1d3ccd2324d)

zugdev commented:
this looks good now! butchered some slow code, I will write a proposal to improve code quality and performance, there is A LOT of redundant stuff here

EresDev commented:
[This one](https://github.com/ubiquity/pay.ubq.fi/pull/383#discussion_r1986701559) is still not resolved

zugdev commented:
## missing permit
When filtering by the provided issue (https://github.com/ubiquity/rpc-handler/issues/60) in Supabase permits table, I only get 4 permits, referring to the first 4 generated:

<img width="762" alt="image" src="https://github.com/user-attachments/assets/5b818dd8-8c18-44a9-9f12-06a6d9348a59" />

Also when filtering by your user ID, I only get one permit with amount = 100*10e18.

<img width="767" alt="image" src="https://github.com/user-attachments/assets/30f42061-c0cc-460a-8bd6-01267f666ca4" />

That permit refers to https://github.com/ubiquity/pay.ubq.fi/issues/379. You can see it's also in your console:
![image](https://github.com/user-attachments/assets/2d4a32d3-8977-4c17-bbbd-6d39b9cf1a5d) it matches the nonce. 

In that case I am inclined to say it was an error in the `text-conversation-rewards` plugin not properly saving it. This further explicits the need of a script that goes through every single issue in our repos and creates the permits in DB from time to time. I have improved the Supabase handling in that script and it's unlikely to happen again, but it's still necessary for the permits generated in the past.

Also notice how the bot created another comment for the rewards when usually it just edits the existing one, might be related.

## loading

Will fix it, though it's not very relevant because of #375

EresDev reviewed (approved):
> In that case I am inclined to say it was an error in the text-conversation-rewards plugin not properly saving it.

Looks like only first permit is stored when we generate multiple permits in issues. To confirm this I wanted to check one more such case and I [randomly checked](https://github.com/ubiquity-os-marketplace/text-conversation-rewards/issues/208) and I also saw that only first permit is stored. So, we never store second permit. it probably made sense at that time because we needed nonce to invalidate the permit and it is same for all permits of an issue. But for this use case, it is a problem because user will get outdated permit. This should be checked further by going through a few more such issues and addressed. 

> This further explicits the need of a script that goes through every single issue in our repos and creates the permits in DB from time to time.

We should improve the way we store them. Going through all issues should be last resort I think.

zugdev commented:
> > In that case I am inclined to say it was an error in the text-conversation-rewards plugin not properly saving it.
> 
> Looks like only first permit is stored when we generate multiple permits in issues. To confirm this I wanted to check one more such case and I [randomly checked](https://github.com/ubiquity-os-marketplace/text-conversation-rewards/issues/208) and I also saw that only first permit is stored. So, we never store second permit. it probably made sense at that time because we needed nonce to invalidate the permit and it is same for all permits of an issue. But for this use case, it is a problem because user will get outdated permit. This should be checked further by going through a few more such issues and addressed.
> 
> > This further explicits the need of a script that goes through every single issue in our repos and creates the permits in DB from time to time.
> 
> We should improve the way we store them. Going through all issues should be last resort I think.

I agree I think ultimately it is a viable build - as we basically do that with devpool-directory, but for open issues. There are some instances where permit generation fails completely, and they might be forgotten: https://github.com/ubiquity-os/.github/issues/1 (this one could probably be used to test your hypothesis too). I think an algorithm could be:
1. Get all organizations and repos
2. Get all issues for each repo
3. Filter by closed and solved ones (solved is a state flag defined below)
4. Check for permits (pay.ubq.fi URLs, decode ?claim parameters)
5. Check if signature is in the db, if so mark as solved
6. If it is not in the db, and should be, add the permit and mark as solved
7. If closed & priced and no rewards mark as pending or straight up generate the reward and then go back to step 5

0x4007 reviewed (changes requested):
<img width="763" alt="image" src="https://github.com/user-attachments/assets/662d6994-2660-4b06-83e8-4d1e4bb00aa8" />

When I claimed and switched it displays "view permit" on the wrong permit. It should be checked if its the correct permit.

zugdev commented:
fixed button handling and fixed the infinite divs stacking apparently that wasnt meant to handle pagination

it is important you add the RLS policy so this UI can update the transaction column in permits table, if that's done we can achieve this:

https://github.com/user-attachments/assets/05cb7ced-461d-45e5-993c-e2f0550abcd6

else we just lose context and after you paginate either up or down you lose that transaction hash. I could set up localstorage, but I really think the right way is adding the RLS policy in supabase permits table for updates

zugdev commented:
@0x4007 know ur pretty busy building but would be very unblocking if you can quickly check this out :)

0x4007 commented:
The right way is saving to local storage and then using it as a buffer to sync to Supabase.

0x4007 commented:
One thing that stands out is that I intentionally left like ten unclaimed permits in the last time I used the app but now they are gone. Is it filtering out invalid permits or something?

zugdev commented:
> One thing that stands out is that I intentionally left like ten unclaimed permits in the last time I used the app but now they are gone. Is it filtering out invalid permits or something?

Yes it actively checks the nonce on-chain. Only displays a permit if it's claimable, but still displays if insolvent.

zugdev commented:
> The right way is saving to local storage and then using it as a buffer to sync to Supabase. 
> 
> I'll take a look in a bit. 

I'll save in localstorage too, then.

zugdev commented:
1. Prevent pagination while claiming.: the Erc20 claim code is written on top of `app.reward`. Previously switching pages in the middle of claiming changed `app.reward` and therefore broke it.

https://github.com/user-attachments/assets/2b483b52-b4bd-423c-b51d-67bd7d7aaa2c

2. We save txn hashes in localstorage and attempt to save them in Supabase (which is still not allowed by RLS). Therefore we can smoothly see claimed transaction, even when we paginate or refresh the webapp. About saving from localstorage to Supabase, I did implement a sync function, however I considered it not "safe" nor wise to use it - as the user has access to localstorage he could modify the transaction hash as he wishes. It is also unlikely once we allow `pay.ubq.fi` to update in permits table that `updatePermitsTx()` fails, and if it does localstorage exists as a fallback.

https://github.com/user-attachments/assets/0bdbfe4c-17c2-4a35-8ff2-c788d8d41c44

0x4007 commented:
Sadly I need to wait until I have more unclaimed and valid permits to properly test this!

0x4007 commented:
> We save txn hashes in localstorage

I assumed you also cached because when I reloaded the page it had to slowly reload everything. I guess thats out of scope though. I'm certain there can be way more speed enhancements to the load I don't understand how it can be so slow.

0x4007 commented on static/scripts/rewards/render-transaction/fetch-permits.ts:
```diff
@@ -1,62 +1,91 @@
-    /* empty */
+  // if found no permits
+  if (!permits.length) {
+    setClaimMessage({ type: "Notice", message: `No claim data found.` });
```
You need to run formatting

0x4007 commented:
Looking forward to when it shows the reward source because I accidentally claimed one I wasn't supposed to claim (the travel one was supposed to be time estimated for 1 week not 1 day. https://github.com/ubiquity/business-development/issues/167#issuecomment-2728182068

0x4007 commented:
I think it might be faster to redo the entire app with these points in mind for another task.

0x4007 commented:
@zugdev can you remove the artificial limitation? It seems to be the most annoying part. The other stuff I think we can handle in other tasks.

zugdev commented:
It is not an unnecessary barrier, because of the following function:

```typescript
export function claimErc20PermitHandlerWrapper(app: AppState) {
  return async function claimErc20PermitHandler() {

    const signer = await connectWallet(); // we are re-testing the in-wallet rpc at this point
    if (!signer) {
      buttonController.hideAll();
      toaster.create("error", `Please connect your wallet to claim this reward.`);
      return;
    }

    app.signer = signer; // update this here to be sure it's set if it wasn't before

    buttonController.hideMakeClaim();
    buttonController.showLoader();

    const isPermitClaimable = await checkPermitClaimability(app);
    if (!isPermitClaimable) {
      buttonController.hideLoader();
      return;
    }

    const permit2Contract = new ethers.Contract(permit2Address, permit2Abi, signer);
    if (!permit2Contract) {
      buttonController.hideLoader();
      return;
    }

    const tx = await transferFromPermit(permit2Contract, app.reward);
    if (!tx) {
      buttonController.hideLoader();
      return;
    }

    const receipt = await waitForTransaction(tx, app.reward.networkId);
    if (!receipt) {
      buttonController.hideLoader();
      return;
    }

    // saves in supabase
    const isHashUpdated = await updatePermitTxHash(app.reward.nonce.toString(), receipt.transactionHash);
    if (!isHashUpdated) {
      buttonController.hideLoader();
      return;
    }

    // fetches from supabase
    app.claimTxs = await getClaimedTxs(app);
    getMakeClaimButton().removeEventListener("click", claimErc20PermitHandler);
    buttonController.onlyShowViewClaim();
    toaster.create("success", "Claim Complete.");
  };
}
```

Basically we do a series of async calls using `app.reward`which is the currently displayed reward. If user navigates amidst claiming `app.reward` and app state in general changes. State management in the app needs to be rewritten, I didn't do it here because it's bigger than the scope. Should I do it here anyways?

zugdev commented on .eslintrc:
```diff
@@ -41,7 +41,7 @@
     "@typescript-eslint/naming-convention": [
       "error",
-      { "selector": "interface", "format": ["StrictPascalCase"], "custom": { "regex": "^I[A-Z]", "match": false } },
+      { "selector": "interface", "format": ["PascalCase"], "custom": { "regex": "^I[A-Z]", "match": false } },
```
don't remember, will revert

zugdev replied:
<img width="442" alt="image" src="https://github.com/user-attachments/assets/cac03901-5ae3-4491-9b37-2d973802575e" />

because of `OAuth`apparently
